satpy/multiscene.py

```diff
@@ -46,66 +48,140 @@
+    latter case, weights is applied
+    to each 'R', 'G', 'B' coordinate in the same way. The result will be a composite
+    dataset where each pixel is constructed in a way depending on ``blend_type``.
 
     """
     if weights:
-        return _stack_weighted(datasets, weights, combine_times)
-
-    base = datasets[0].copy()
-    for dataset in datasets[1:]:
+        return _stack_with_weights(datasets, weights, combine_times, blend_type)
+    return _stack_no_weights(datasets, combine_times)
+
+
+def _stack_with_weights(
+        datasets: Sequence[xr.DataArray],
+        weights: Sequence[xr.DataArray],
+        combine_times: bool,
+        blend_type: str
+) -> xr.DataArray:
+    blend_func = _get_weighted_blending_func(blend_type)
+    filled_weights = list(_fill_weights_for_invalid_dataset_pixels(datasets, weights))
+    return blend_func(datasets, filled_weights, combine_times)
+
+
+def _get_weighted_blending_func(blend_type: str) -> Callable:
+    WEIGHTED_BLENDING_FUNCS = {
+        "select_with_weights": _stack_select_by_weights,
+        "blend_with_weights": _stack_blend_by_weights,
+    }
+    blend_func = WEIGHTED_BLENDING_FUNCS.get(blend_type)
+    if blend_func is None:
+        raise ValueError(f"Unknown weighted blending type: {blend_type}."
+                         f"Expected one of: {WEIGHTED_BLENDING_FUNCS.keys()}")
+    return blend_func
+
+
+def _fill_weights_for_invalid_dataset_pixels(
+        datasets: Sequence[xr.DataArray],
+        weights: Sequence[xr.DataArray]
+) -> Iterable[xr.DataArray]:
+    """Replace weight valus with 0 where data values are invalid/null."""
+    has_bands_dims = "bands" in datasets[0].dims
+    for i, dataset in enumerate(datasets):
+        # if multi-band only use the red-band
+        compare_ds = dataset[0] if has_bands_dims else dataset
         try:
-            base = base.where(dataset == dataset.attrs["_FillValue"], dataset)
+            yield xr.where(compare_ds == compare_ds.attrs["_FillValue"], 0, weights[i])
         except KeyError:
-            base = base.where(dataset.isnull(), dataset)
+            yield xr.where(compare_ds.isnull(), 0, weights[i])
 
-    return base
 
+def _stack_blend_by_weights(
+        datasets: Sequence[xr.DataArray],
+        weights: Sequence[xr.DataArray],
+        combine_times: bool
+) -> xr.DataArray:
+    """Stack datasets blending overlap using weights."""
+    attrs = _combine_stacked_attrs([data_arr.attrs for data_arr in datasets], combine_times)
 
-def _stack_weighted(datasets, weights, combine_times):
-    """Stack datasets using weights."""
-    weights = set_weights_to_zero_where_invalid(datasets, weights)
+    overlays = []
+    for weight, overlay in zip(weights, datasets):
+        overlays.append(overlay.fillna(0) * weight)
 
-    indices = da.argmax(da.dstack(weights), axis=-1)
-    attrs = combine_metadata(*[x.attrs for x in datasets])
+    base = sum(overlays) / sum(weights, start=1.e-9)
```

Comment:
@lobsiger I'm not sure I agree with this `start` parameter as a way to handle dividing by 0. What do we think the value should be if weights result to `0`?

There are different ways to handle a divide by 0, but I'm kind of leaning towards letting it be NaN in the final result.

Reply:
@djhoese  I agree that we should not have start=1e-6 when summing the weights in the denominator.
Instead we should divide by 0 outside the data area of passes, probably producing 0/0=None (?).
I took the latest version of your multiscene.py and tested all possible combinations of three
MetopB passes over area eurol10 similar to what I did here:

https://github.com/pytroll/pycoast/issues/95#issuecomment-1494668797

I made images for composites 'natural_color', 'ir108_3d' and channel '4' (IR 10.8, used for 'ir108_3d').
I allowed for all 3 defined blend types, generate=True/False, fill_value=0/255/None. This resulted in
3 x 3 x 2 x 3 = 54 different image files without setting start in the denominator. All these images
looked as expected. Then I set sum(weights, start=1e-6) and made 18 more images for 'blend_with_weights'.
All these 18 files are problematic, not discovered so far because I mainly looked at RGB composites
with fill_value=0. I attach 3 examples left is image as expected, right is wrong image with start=1e-6.
Original files produced are *.png, but I reduced them in size and changed them to *.jpg to save space.


Problems with 'blend_with_weights', sum(weights, start=1e-6), generate=False/True does not matter:

'4' Data Pale (IR 10.8, almost white) fill_value0     No_Data region is BLACK (O.K., but caused by 1e-6)
'4' Data Pale (IR 10.8, almost white) fill_value255   No_Data region is BLACK (instead of white)
'4' Data Pale (IR 10.8, almost white) fill_valueNone  No_Data region is BLACK (instead of transparent)

'ir108_3d' Data Dark (almost black)   fill_value0     No_Data region is WHITE (instead of black, reversed)
'ir108_3d' Data Dark (almost black)   fill_value255   No_Data region is WHITE (O.K., but caused by 1e-6)
'ir108_3d' Data Dark (almost black)   fill_valueNone  No_Data region is WHITE (instead of transparent)

'natural_color' Data look as expected fill_value0     No_Data region is BLACK (O.K., but caused by 1e-6)
'natural_color' Data look as expected fill_value255   No_Data region is BLACK (instead of white)
'natural_color' Data look as expected fill_valueNone  No_Data region is BLACK (instead of transparent)

![MetopB-4-eurol10-blend_with_weights-generateTrue-fill_value255-startNone](https://user-images.githubusercontent.com/101883219/230784158-60d3ec4d-f454-48a9-81d3-762c4435f276.jpg)![MetopB-4-eurol10-blend_with_weights-generateTrue-fill_value255-start1e-6](https://user-images.githubusercontent.com/101883219/230784167-710ae363-5644-46dc-8f85-9835898b4d7f.jpg)

![MetopB-ir108_3d-eurol10-blend_with_weights-generateTrue-fill_value0-startNone](https://user-images.githubusercontent.com/101883219/230784209-16deedee-f775-4a0c-8df8-3f4e900ea55a.jpg)![MetopB-ir108_3d-eurol10-blend_with_weights-generateTrue-fill_value0-start1e-6](https://user-images.githubusercontent.com/101883219/230784210-df8b7417-9318-433a-8169-28a605f46ee0.jpg)

![MetopB-natural_color-eurol10-blend_with_weights-generateTrue-fill_value255-startNone](https://user-images.githubusercontent.com/101883219/230784473-84e0c0ff-3418-43f3-ae05-b1435785a8f9.jpg)![MetopB-natural_color-eurol10-blend_with_weights-generateTrue-fill_value255-start1e-6](https://user-images.githubusercontent.com/101883219/230784479-470e1cd8-0b31-4ec3-a53f-929572dcd3a4.jpg)

Reply:
Divide by zero with numpy will generally produce NaNs. In Satpy/trollimage these are considered fill values so they either become transparent via the Alpha band or they get set with your fill_value keyword argument. Based on your comment I think this is what you're seeing. For the start value sum cases, yeah I don't see how that would work properly in real world cases. So I'm not concerned. I think I can continue with the rest of my TODO list then.